File: src/main/java/com/advent2021j/puzzle2/Puzzle2.java
package com.advent2021j.puzzle2;

import com.advent2021j.base.BaseJ;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

enum Direction { forward, up, down }
class InputLine {
    Direction dir;
    int x;

    public InputLine(Direction dir, int x) {
        this.dir = dir;
        this.x = x;
    }
}

class Position {
    int depth = 0;
    int distance = 0;
    int aim = 0;

    public Position() {
    }

    public Position(int depth, int distance, int aim) {
        this.depth = depth;
        this.distance = distance;
        this.aim = aim;
    }

    int solution() { return depth * distance; }

}

public class Puzzle2 extends BaseJ<List<InputLine>, Integer, Integer> {
    public static void main(String[] args) {
        try {
            Puzzle2 puz = new Puzzle2();
            Integer solution1 = puz.solvePuzzle("inputs.txt", new ArrayList<>());
            System.out.println("Solution1: " + solution1);

            Integer solution2 = puz.solvePuzzle2("inputs.txt", new ArrayList<>());
            System.out.println("Solution2: " + solution2);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void parseLine(String line, List<InputLine> data) {
         String[] parts = line.split(" ");
         Direction dir = Direction.valueOf(parts[0]);
         int x = Integer.parseInt(parts[1]);
         data.add(new InputLine(dir, x));
    }

    @Override
    protected Integer computeSolution(List<InputLine> data) {
        Position pos = new Position();
        for (InputLine datum : data) {
            switch (datum.dir) {
                case forward:
                    pos = new Position(pos.depth, pos.distance + datum.x, pos.aim);
                    break;
                case up:
                    int x = -datum.x;
                    pos = new Position(pos.depth + x, pos.distance, pos.aim);
                    break;
                case down:
                    pos = new Position(pos.depth + datum.x, pos.distance, pos.aim);
                    break;
            }
        }
        return pos.solution();
    }

    @Override
    protected Integer computeSolution2(List<InputLine> data) {
        Position pos = new Position();
        for (InputLine datum : data) {
            switch (datum.dir) {
                case forward:
                    pos = new Position(pos.depth, pos.distance + datum.x, pos.aim);
                    pos = new Position(pos.depth + pos.aim * datum.x, pos.distance, pos.aim);
                    break;
                case up:
                    int x = -datum.x;
                    pos = new Position(pos.depth, pos.distance, pos.aim + x);
                    break;
                case down:
                    pos = new Position(pos.depth, pos.distance, pos.aim + datum.x);
                    break;
            }
        }
        return pos.solution();
    }
}

File: src/main/kotlin/com/advent2021/puzzle2/Puzzle2.kt
package com.advent2021.puzzle2

import com.advent2021.base.Base
import com.advent2021.puzzle2.Position.Companion.incDepth
import com.advent2021.puzzle2.Position.Companion.incDistance
import com.advent2021.puzzle2.Position2.Companion.incAim

enum class Direction { forward, up, down }
data class InputLine(
    val dir: Direction,
    val x: Int
)
typealias Data = ArrayList<InputLine>
typealias Solution = Int
typealias Solution2 = Int

fun main() {
    try {
        val puz = Puzzle2()
        val solution1 = puz.solvePuzzle("inputs.txt", Data())
        println("Solution1: $solution1")

        val solution2 = puz.solvePuzzle2("inputs.txt", Data())
        println("Solution2: $solution2")
    } catch (t: Throwable) {
        t.printStackTrace()
    }
}

open class Position(
    var depth: Int = 0,
    var distance: Int = 0
) {
    val solution: Int
        get() { return depth * distance }

    companion object {
        fun <Pos: Position> incDistance(x: Int, pos: Pos): Pos = pos.also { pos.distance += x }
        fun <Pos: Position> incDepth(x: Int, pos: Pos): Pos = pos.also { pos.depth += x }
    }
}

class Position2(
    depth: Int = 0,
    distance: Int = 0,
    var aim: Int = 0
) : Position(depth, distance) {
    companion object {
        fun incAim(x: Int, pos: Position2): Position2 = pos.also { pos.aim += x }
    }
}


class Puzzle2 : Base<Data, Solution, Solution2>() {
    override fun parseLine(line: String, data: Data) {
        val parts = line.split(" ")
        data.add(InputLine(Direction.valueOf(parts[0]), parts[1].toInt()))
    }

    override fun computeSolution(data: Data): Solution {
        return with(data.fold(Position()) { acc, line ->
            when (line.dir) {
                Direction.forward -> incDistance(line.x, acc)
                Direction.down -> incDepth(line.x, acc)
                Direction.up -> incDepth(-line.x, acc)
            }
        }) { solution }
    }

    override fun computeSolution2(data: Data): Solution2 {
        return with(data.fold(Position2()) { acc, line ->
            when (line.dir) {
                Direction.forward -> incDistance(line.x, acc).also { incDepth(acc.aim * line.x, it) }
                Direction.down -> incAim(line.x, acc)
                Direction.up -> incAim(-line.x, acc)
            }
        }) { solution }
    }
}

